core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -153,13 +141,23 @@ public boolean tryComplete() {
             shareFetchData.groupId(), shareFetchData.memberId(),
             shareFetchData.partitionMaxBytes().keySet());
 
-        topicPartitionDataFromTryComplete = acquirablePartitions();
+        Map<TopicIdPartition, FetchRequest.PartitionData> topicPartitionData = acquirablePartitions();
 
-        if (!topicPartitionDataFromTryComplete.isEmpty())
-            return forceComplete();
-        log.info("Can't acquire records for any partition in the share fetch request for group {}, member {}, " +
-                "topic partitions {}", shareFetchData.groupId(),
-                shareFetchData.memberId(), shareFetchData.partitionMaxBytes().keySet());
+        if (!topicPartitionData.isEmpty()) {
+            maybeUpdateFetchOffsetMetadataForTopicPartitions(topicPartitionData);
```

Comment:
It's possible that `maybeUpdateFetchOffsetMetadataForTopicPartitions` calls `readFromLog`, which returns enough bytes. In that case, it's more efficient to reuse the fetched result instead of calling `readFromLog` again in `onComplete`.